### PR TITLE
Story8

### LINKED ISSUE #146: Include weather data
- [x] Look into MET API
- [x] Find an alternative to historical data, if needed
- [ ] Decide on how to incorporate weather API
- [ ] XML to JSON
- [ ] SoonTM

 More will follow

Possible improvements:

Edge cases:

### PR DESCRIPTION
**Pull request by Pontius1007**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Description

Weather should now work to a certain extent. 

## Motivation and Context

Story 8, resolves #146 

## How Has This Been Tested?

We have some tests written by sebastian, but we could really use some manual testing/breaking before we go into prod. Remember this is written with time as a limiting factor and the code might reflect that. Please QA in-depth. 

Check:
Add weather data.
Add prod data to same segment and within the weather timeframe. See if it updates
Add weather data with a timeframe that includes a prod data. The weather should not be added then
+++ 

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: apps/api/tests/test_weather.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    def test_for_correct_segment(self):
+        """
+        Check that the road segment has the correct weather data
+        """
+        segment = RoadSegment.objects.get()
+        weather = WeatherData.objects.get()
+        self.assertEqual(segment.id, weather.segment.id)
+
+    def test_for_updated_value(self):
```
Comment (flamberhamber) [inline apps/api/tests/test_weather.py]: mangler docstring

File: apps/api/tests/test_weather.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            plow_active=True, wet_spreader_active=True, brush_active=True, material_type_code=True,
+            segment=RoadSegment.objects.get()
+        )
+        weather.map_weather_to_segment([{"start_time_period": '2018-12-10T08:45:15Z',
+                                         "end_time_period": '2018-12-11T08:45:15Z', "county_and_municipality_id": 5001,
+                                         "value": 4, "unit": 'mm', "degrees": 30, "segment": 4}])
+        entry = WeatherData.objects.get()
+        self.assertEqual(entry.value, 2)
+
+    def test_for_existing_weather_data(self):
```
Comment (flamberhamber) [inline apps/api/tests/test_weather.py]: er ingen sjekk her

File: apps/api/weather/weather.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+def check_for_existing_data(entry):
+    """
+    :param entry: JSON weather object
+    :return: True or False if the same road segment has production data within the time range for the weather
+    """
+    start = entry['start_time_period']
+    end = entry['end_time_period']
+    segment = entry['segment_id']
+    prod_data_list = ProductionData.objects.filter(segment=segment)
+    for data in prod_data_list:
```
Comment (flamberhamber) [inline apps/api/weather/weather.py]: Vil den ikke returnere etter å sjekke det første elementet? Tror det er redundant med en loop her

File: apps/api/weather/weather.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        return data.time <= end and data.time >= start
+    return False
```
Comment (flamberhamber) [inline apps/api/weather/weather.py]: Return False vil aldri inntreffe

**Comment by sardred** [comment]
`End time` som er før `start time` burde utløse egen feilmelding
Kan ha negativ county/municipality ID
Kan sende inn negativ ID
partial_update: No fields are mandatory, men id er oblig
403 på delete som superuser

**Comment by haavahe** [comment]
Er ikke mulig å inputte start eller sluttid med tidssone: 2016-11-04T08:45:15+02:00.

File: apps/api/serializers.py [hunk 5]
```diff
@@ -29,6 +29,29 @@ class Meta:
...
+    Serializer for validating the weather data input. Does not need the save and update methods.
+    """
+    start_time_period = serializers.DateTimeField(help_text="Start time for the weather period. "
+                                                            "Example: 2018-12-09T08:45:15")
+    end_time_period = serializers.DateTimeField(help_text="End time for the weather period. "
+                                                          "Example: 2018-12-10T08:45:15")
+    county_and_municipality_id = serializers.IntegerField(help_text="County and municipality number put together."
+                                                                    "Example: 5001 for Trondheim")
+    value = serializers.IntegerField(help_text="The amount of precipitation without the unit. Example: 2")
+    unit = serializers.CharField(max_length=2, help_text="The unit describing the value. Max two in length."
```
[the remark this answers is not in the document]
Comment (Pontius1007) [reply apps/api/serializers.py]: Kan vel bare sette default til 'mm' og ikke la brukeren skrive det inn upside_down_face

File: apps/api/serializers.py [hunk 6]
```diff
@@ -29,6 +32,58 @@ class Meta:
...
+            raise serializers.ValidationError("End time can not be before start time")
+        elif (timezone.now() - data['end_time_period']) > timedelta(days=1):
+            raise serializers.ValidationError("Weather can not be over 24 hours old")
+        elif data['end_time_period'] > timezone.now():
+            raise serializers.ValidationError("End time can not be in the future")
+        elif (data['end_time_period'] - data['start_time_period']) >= timedelta(days=1):
+            raise serializers.ValidationError("Only supports 1 day time frame for weather")
+        return data
+
+    start_time_period = serializers.DateTimeField(help_text="Start time for the weather period. "
```
Comment (haavahe) [inline apps/api/serializers.py]: Kan du putte disse feltene over validatemetodene? Det gjør vondt å se på

File: apps/api/views.py [hunk 7]
```diff
@@ -123,6 +127,66 @@ def create(self, request, *args, **kwargs):
...
+        # Map weather data to road a road segment
+        number_of_updated_weather, mapped_weather = weather.map_weather_to_segment(data)
+
+        # Instantiate the serializer
+        serializer = WeatherDataSerializer(data=mapped_weather, many=True)
+
+        # Check if the serializer is valid and takes the necessary actions
+        if serializer.is_valid(raise_exception=True):
+            serializer.save()
+            # headers = self.get_success_headers(serializer.data)
```
Comment (haavahe) [inline apps/api/views.py]: Remove pls

**Pull request merged**